AuthSAML.php

```diff
@@ -131,125 +152,175 @@ public function beforeLogin() {
         }
     }
 
+    /**
+     * @see event
+     * Redirect to the logout set if needed
+     **/
     public function afterLogout()
     {
         $ssp = $this->get_saml_instance();
+        if (is_null($ssp)) {
+            return;
+        }
         $redirect = $this->get('logout_redirect', null, null, '/admin');
         if ($ssp->isAuthenticated()) {
             Yii::app()->controller->redirect($ssp->getLogoutUrl($redirect));
             Yii::app()->end();
         }
     }
 
+    /**
+     * @see event
+     * Add the content needed for SAML login in login form
+     **/
     public function newLoginForm()
     {
+        $newLoginFormEvent = $this->getEvent();
         $authtype_base = $this->get('authtype_base', null, null, 'Authdb');
-
+        if (App()->getSession()->get('auth_saml_invalid')) {
+            $invalidAccount = $this->gT('Invalid SAML account for this service');
+            $content = '<div class="alert alert-warning">' . $invalidAccount . '</div>';
+            $newLoginFormEvent->getContent($authtype_base)->addContent($content, 'prepend');
+            return;
+        }
         $ssp = $this->get_saml_instance();
-        $this->getEvent()->getContent($authtype_base)->addContent('<li><center>Click on that button to initiate SAML Login<br><a href="'.$ssp->getLoginURL().'" title="SAML Login"><img src="'.Yii::app()->getConfig('imageurl').'/saml_logo.gif"></a></center><br></li>', 'prepend');
+        if (is_null($ssp)) {
+            return;
+        }
+        $description = $this->gT('Click on that button to initiate SAML Login');
+        $imageUrl = Yii::app()->getConfig('imageurl') . '/saml_logo.gif';
+        $imageAlt = $this->gT('SAML Login');
+        /* No usage of twig : LimeSurvey twig replaced by simplesamlphp due to autoload
+         * Break on limesurvey 3 and 5
+         * @todo : find a way to fix it, or use twig only for 6
+         **/
+        $content = '<div class="saml-container">' .
+                    '<div class="text-center">' . $description . '</div>' .
+                    '<div class="text-center">' .
+                    '<a href="' . $ssp->getLoginURL() . '" title="' . $imageAlt . '" class="btn btn-primary">' .
+                    '<img src="' . $imageUrl . '" alt="' . $imageAlt . '">' .
+                    '</a>' .
+                    '</div>' .
+                    '</div>';
+        $newLoginFormEvent->getContent($authtype_base)->addContent($content, 'prepend');
     }
 
-    public function newUserSession() {
-
+    /**
+     * @see event
+     * Create or update user if authneticated by SAML
+     **/
+    public function newUserSession()
+    {
+        $oEvent = $this->getEvent();
+        $identity = $oEvent->get('identity');
+        if ($identity->plugin != 'AuthSAML') {
+            return;
```

Comment:
Here, this line prevents users from logging in using local accounts. If i comment it out then it works fine both with local and SAML accounts

Reply:
But ? If user came from AuthDB : other action must not be done !

Like AuthLDAP for example https://github.com/LimeSurvey/LimeSurvey/blob/890835d8e21048850b80ce6e3a09aa0837f65b10/application/core/plugins/AuthLDAP/AuthLDAP.php#L436

Reply:
@vpanag : please : lets continue on https://github.com/Shnoulle/Limesurvey-SAML-Authentication/issues/1